Do H2D/D2H of input/result on separate threads/cuda.Streams

mikaylagawarecki opened the pull request:
Added two `ThreadPoolExecutor`s with 1 worker each for D2H and H2D copies. Each uses its own `cuda.Stream`. The purpose is to try to overlap D2H and H2D with compute and allow the worker handling prediction to launch compute kernels without being blocked by D2H/H2D.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116190
* __->__ #116189
* #116188
* #116187
* #115286

mikaylagawarecki committed:
Do H2D/D2H of input/result on separate threads/cuda.Streams

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Do H2D/D2H of input/result on separate threads/cuda.Streams"

Added two `ThreadPoolExecutor`s with 1 worker each for D2H and H2D copies. Each uses its own `cuda.Stream`. The purpose is to try to overlap D2H and H2D with compute and allow the worker handling prediction to launch compute kernels without being blocked by D2H/H2D.




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Do H2D/D2H of input/result on separate threads/cuda.Streams"

Added two `ThreadPoolExecutor`s with 1 worker each for D2H and H2D copies. Each uses its own `cuda.Stream`. The purpose is to try to overlap D2H and H2D with compute and allow the worker handling prediction to launch compute kernels without being blocked by D2H/H2D.




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Do H2D/D2H of input/result on separate threads/cuda.Streams"

Added two `ThreadPoolExecutor`s with 1 worker each for D2H and H2D copies. Each uses its own `cuda.Stream`. The purpose is to try to overlap D2H and H2D with compute and allow the worker handling prediction to launch compute kernels without being blocked by D2H/H2D.




[ghstack-poisoned]

vadimkantorov commented:
Maybe sth like this can be built-in in a new copy wrapping function, such as : https://github.com/pytorch/pytorch/issues/69431

which could be reused by user code

albanD commented on benchmarks/inference/server.py:
```diff
@@ -207,14 +210,45 @@ def _setup(self):
+    def copy_data(self, input_buffer, data, copy_event, copy_sem):
+        data = data.pin_memory()
+        with torch.cuda.stream(self.h2d_stream):
+            input_buffer.copy_(data, non_blocking=True)
```
What happens if you make this non_blocking=False?
So that the copying thread waits for the copy to be done instead of the thread handling model forward maybe waiting for the copy in to be finished?

mikaylagawarecki replied:
I was thinking to have this be `non_blocking=True` so that the copying thread can overlap the next `pin_memory()` with the current H2D copy, would that make sense?

albanD replied:
It does, but it also means that the main worker thread (that forwards the model) might be called and stuck waiting for the copy to finish before being able to forward.
And I think that blocking the main worker thread is the one thing we really don't want to see happen.

mikaylagawarecki replied:
Hmm I'm not sure I follow, I think the predictor thread (which is separate from the main thread) has to be stuck and wait for the copy

In this case, there is a `copy_semaphore` and `copy_event` for each input request, so we would have something like

```python
# In copying thread
input1.to(cuda, non_blocking=True)
# input_sem1 makes sure copy_event1.record() is called
# copy_event1 would wait for the H2D to complete
# Next kernel issue for H2D for input2 is not blocked in copy thread
input2.to(cuda, non_blocking = True)
# input_sem2 makes sure copy_event2.record() is called

# In predictor thread
# acquire sem1
# wait for copy_event1 (this ensures that the predictor thread's GPU kernels see the correct input data)
model.forward(input)
```


In the multi-thread case (next PR) because there are multiple predictor threads `input_sem1`/`input_event1` would not be blocking the predictor thread handling `input2` (if `input1` and `input2` are assigned to different predictor threads in the pool)

albanD replied:
My point is that having "wait for copy_event1" sounds bad!
This predictor thread is the "hot path" for the whole system and so should never be waiting.

mikaylagawarecki replied:
Doesn't the predictor thread have to wait for the input to be copied before it can run forward?

In my understanding, we need to ensure the above dependency is respected:
- Current (`copy_data` with `non_blocking=True`):
    - The main thread manages the thread pools for the predictor thread and copy threads. I think we do not block the main thread on any copies! 😄 
    - For CPU correctness, the semaphore ensures that the predictor thread does not run until the copy thread has issued the non-blocking H2D copy and recorded its copy event.
    - For GPU correctness, the predictor thread has its stream wait on the copy event to ensure model forward kernels see the copied input data.
- Alternative (`copy_data` with `non_blocking=False`):
    - The difference is that the copy thread waits for the H2D copy to finish before releasing the semaphore, meaning that the predictor thread does not run any real instructions (e.g. issuing model forward kernels) until the H2D copy has finished.

albanD replied:
> Doesn't the predictor thread have to wait for the input to be copied before it can run forward?

It has to, but that waiting can be done by the other thread while the predictor thread is handling another query.

> meaning that the predictor thread does not run any real instructions (e.g. CPU issuing model forward GPU kernels that will wait on the copy_event) 

Ho interesting, so this would mean that for low batch-size/gpu usage, this is going to increase latency by reducing overlap. While for high gpu use case, it should improve throughput by allowing the predictor thread that is the bottleneck (hopefully) to never wait on anything.

mikaylagawarecki replied:
Hmmmm is the blocking thing that you are concerned about the semaphore or the cuda.Event? In any case, I don't fully understand why the predictor thread shouldn't wait for its own input copy to be issued, but I will try it with `non_blocking=False` and compare the profiles to understand

mikaylagawarecki committed:
Update on "Do H2D/D2H of input/result on separate threads/cuda.Streams"

Added two `ThreadPoolExecutor`s with 1 worker each for D2H and H2D copies. Each uses its own `cuda.Stream`. The purpose is to try to overlap D2H and H2D with compute and allow the worker handling prediction to launch compute kernels without being blocked by D2H/H2D.




[ghstack-poisoned]
